File: 30_day_challenge_2020_December/941_valid_mountain_array.cpp
/****************************************************************
* Author            : yiorgosynkl (find me in Github: https://github.com/yiorgosynkl)
* Date created      : 20201210
* Problem link      : https://leetcode.com/problems/valid-mountain-array/
****************************************************************/

class Solution {
public:
// bool validMountainArray(vector<int>& arr) {
//     int n = arr.size();
//     int i = 0, j = n - 1;
//     for (; i + 1 < n; i++ ){
//         if (arr[i] >= arr[i+1])
//             break;
//     }
//     for (; j > 0; j-- ){
//         if (arr[j-1] <= arr[j])
//             break;
//     }        
//     return (n >= 3 && 0 < i && i == j && j < n-1);
// }

// @lee215
bool validMountainArray(vector<int>& A) {
    int n = A.size(), i = 0, j = n - 1;
    while (i + 1 < n && A[i] < A[i + 1]) i++;
    while (j > 0 && A[j - 1] > A[j]) j--;
    return i > 0 && i == j && j < n - 1;
}

};

File: 30_day_challenge_2020_December/941_valid_mountain_array.py
################################################################
# Author            : yiorgosynkl (find me in Github: https://github.com/yiorgosynkl)
# Date created      : 20201210
# Problem link      : https://leetcode.com/problems/valid-mountain-array/
################################################################

class Solution:
    # def validMountainArray(self, arr: List[int]) -> bool:
    #     if len(arr) < 3: return False
    #     i, ln = 0, len(arr)
    #     while i < ln-1 and arr[i] < arr[i+1]: i += 1
    #     if i == 0 or i == ln-1: return False
    #     while i < ln-1 and arr[i] > arr[i+1]: i += 1
    #     return i == ln-1
    
    # @lee215
    def validMountainArray(self, A):
        i, j, n = 0, len(A) - 1, len(A)
        while i + 1 < n and A[i] < A[i + 1]: i += 1
        while j > 0 and A[j - 1] > A[j]: j -= 1
        return 0 < i == j < n - 1